src/backend/distributed/utils/colocation_utils.c

```diff
@@ -0,0 +1,213 @@
+		return false;
+	}
+
+	return leftColocationId == rightColocationId;
+}
+
+
+/*
+ * ShardsColocated function checks whether given two shards are co-located and
+ * returns true if they are co-located. Two shards are co-located only if;
+ * - Tables containing the shards are co-located.
+ * - Min/Max values of the shards are same.
+ */
+bool
+ShardsColocated(ShardInterval *leftShardInterval, ShardInterval *rightShardInterval)
+{
+	bool tablesColocated = TablesColocated(leftShardInterval->relationId,
+										   rightShardInterval->relationId);
+
+	if (tablesColocated)
+	{
+		/*
+		 * Having co-located tables implies that tables are partitioned by hash partition
+		 * therefore it is safe to use DatumGetInt32 here.
+		 */
+		int leftShardMinValue = DatumGetInt32(leftShardInterval->minValue);
+		int leftShardMaxValue = DatumGetInt32(leftShardInterval->maxValue);
+		int rightShardMinValue = DatumGetInt32(rightShardInterval->minValue);
+		int rightShardMaxValue = DatumGetInt32(rightShardInterval->maxValue);
+
+		bool minValuesEqual = leftShardMinValue == rightShardMinValue;
+		bool maxValuesEqual = leftShardMaxValue == rightShardMaxValue;
+
+		return minValuesEqual && maxValuesEqual;
+	}
+
+	return false;
+}
+
+
+/*
+ * ColocatedTableList function returns list of relation ids which are co-located
+ * with given table. If given table is not hash distributed, co-location is not
+ * valid for that table and it is only co-located with itself.
+ */
+List *
+ColocatedTableList(Oid distributedTableId)
+{
+	int tableColocationId = TableColocationId(distributedTableId);
+	List *colocatedTableList = NIL;
+
+	Relation pgDistPartition = NULL;
+	TupleDesc tupleDescriptor = NULL;
+	SysScanDesc scanDescriptor = NULL;
+	HeapTuple heapTuple = NULL;
+	bool indexOK = true;
+	int scanKeyCount = 1;
+	ScanKeyData scanKey[1];
+
+	/*
+	 * If distribution type of the table is not hash, the table is only co-located
+	 * with itself.
+	 */
+	if (tableColocationId == INVALID_COLOCATION_ID)
+	{
+		colocatedTableList = lappend_oid(colocatedTableList, distributedTableId);
+		return colocatedTableList;
+	}
+
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_partition_colocationid,
+				BTEqualStrategyNumber, F_INT8EQ, ObjectIdGetDatum(tableColocationId));
+
+	pgDistPartition = heap_open(DistPartitionRelationId(), AccessShareLock);
+	tupleDescriptor = RelationGetDescr(pgDistPartition);
+	scanDescriptor = systable_beginscan(pgDistPartition,
+										DistPartitionColocationidIndexId(),
+										indexOK, NULL, scanKeyCount, scanKey);
+
+	heapTuple = systable_getnext(scanDescriptor);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		bool isNull = false;
+		Oid colocatedTableId = heap_getattr(heapTuple,
+											Anum_pg_dist_partition_logicalrelid,
+											tupleDescriptor, &isNull);
+
+		colocatedTableList = lappend_oid(colocatedTableList, colocatedTableId);
+		heapTuple = systable_getnext(scanDescriptor);
+	}
+
+	systable_endscan(scanDescriptor);
+	heap_close(pgDistPartition, NoLock);
```

Comment:
Why we close with NoLock? On `LookupDistShardTuples` we have the same pattern with `AccessShareLock` on the close.